### PR TITLE
PLANNER-1819 Add Examination CSProvider

### PR DESCRIPTION
**Pull request by michaltomco**
Add a constraint stream provider for the Examination example.

### CODE REVIEW

**Comment by michaltomco** [comment]
The constraints at the moment are rather slow, not sure whether they are properly converted from the Drools language. I've used a lot of filtering after noticing it is not supported yet, so I've put .filter() instead.

A few questions:
- What is the ideal way how to model the != null? Two options come to mind, either use .ifNotExists() or .filter(!=null). What is the ideal one performance-wise?
- Regarding roomCapacityTooSmall() constraint, is the drools accumulate function properly converted? If not what is the ideal way to do this?

File: optaplanner-examples/src/main/java/org/optaplanner/examples/examination/optional/score/ExaminationConstraintProvider.java [hunk 1]
```diff
@@ -0,0 +1,371 @@
...
+    private Constraint periodPenaltyExamCoincidence(ConstraintFactory constraintFactory) {
+        return constraintFactory.from(PeriodPenalty.class)
+                .filter(periodPenalty -> periodPenalty.getPeriodPenaltyType() == PeriodPenaltyType.EXAM_COINCIDENCE)
+                .join(Exam.class,
+                      equal(PeriodPenalty::getLeftTopic, Exam::getTopic))
+                .filter((periodPenalty, leftExam) -> leftExam.getPeriod() != null)
+                .join(Exam.class,
+                      equal((periodPenalty, exam) -> periodPenalty.getRightTopic(), Exam::getTopic),
+                      equal((periodPenalty, exam) -> exam.getPeriod(), Exam::getPeriod))
+                .filter((periodPenalty, leftExam, rightExam) -> rightExam.getPeriod() != null)
```
Comment (triceo) [inline optaplanner-examples/src/main/java/org/optaplanner/examples/examination/optional/score/ExaminationConstraintProvider.java]: This is no longer necessary, right. `leftExam.getPeriod()` already isn't `null`, and `rightExam.getPeriod()` is joined where equal.

File: optaplanner-examples/src/main/java/org/optaplanner/examples/examination/optional/score/ExaminationConstraintProvider.java [hunk 2]
```diff
@@ -0,0 +1,371 @@
...
+    private Constraint periodPenaltyAfter(ConstraintFactory constraintFactory) {
+        return constraintFactory.from(PeriodPenalty.class)
+                .filter(periodPenalty -> periodPenalty.getPeriodPenaltyType() == PeriodPenaltyType.AFTER)
+                .join(Exam.class,
+                      equal(PeriodPenalty::getLeftTopic, Exam::getTopic))
+                .filter((periodPenalty, leftExam) -> leftExam.getPeriod() != null)
+                .join(Exam.class,
+                      equal((periodPenalty, leftExam) -> periodPenalty.getRightTopic(), Exam::getTopic),
+                      lessThanOrEqual((periodPenalty, leftExam) -> leftExam.getPeriodIndex(), Exam::getPeriodIndex))
+                .filter((periodPenalty, leftExam, rightExam) -> rightExam.getPeriod() != null)
```
Comment (triceo) [inline optaplanner-examples/src/main/java/org/optaplanner/examples/examination/optional/score/ExaminationConstraintProvider.java]: Can  `getPeriod()` be `null` if `getPeriodIndex()` already worked fine?

File: optaplanner-examples/src/main/java/org/optaplanner/examples/examination/optional/score/ExaminationConstraintProvider.java [hunk 3]
```diff
@@ -0,0 +1,371 @@
...
+    private Constraint roomPenaltyExclusive(ConstraintFactory constraintFactory) {
+        return constraintFactory.from(RoomPenalty.class)
+                .filter(roomPenalty -> roomPenalty.getRoomPenaltyType() == RoomPenaltyType.ROOM_EXCLUSIVE)
+                .join(Exam.class,
+                      equal(RoomPenalty::getTopic, Exam::getTopic))
+                .filter((roomPenalty, leftExam) -> leftExam.getPeriod() != null && leftExam.getRoom() != null)
+                .join(Exam.class,
+                      equal((roomPenalty, leftExam) -> leftExam.getRoom(), Exam::getRoom),
+                      equal((roomPenalty, leftExam) -> leftExam.getPeriod(), Exam::getPeriod))
+                .filter((roomPenalty, leftExam, rightExam) -> leftExam.getTopic().equals(rightExam.getTopic()))
```
Comment (triceo) [inline optaplanner-examples/src/main/java/org/optaplanner/examples/examination/optional/score/ExaminationConstraintProvider.java]: Can't you do this using `Joiners.equal()` within the preceding `join()`?

File: optaplanner-examples/src/main/java/org/optaplanner/examples/examination/optional/score/ExaminationConstraintProvider.java [hunk 4]
```diff
@@ -0,0 +1,371 @@
...
+//    end
+    private Constraint periodSpread(ConstraintFactory constraintFactory) {
+        return constraintFactory.from(TopicConflict.class)
+                .join(Exam.class,
+                      equal(TopicConflict::getLeftTopic, Exam::getTopic))
+                .filter((topicConflict, leftExam) -> leftExam.getPeriod() != null)
+                .join(Exam.class,
+                      equal((topicConflict, leftExam) -> topicConflict.getRightTopic(), Exam::getTopic),
+                      equal((topicConflict, leftExam) -> leftExam.getDayIndex(), Exam::getDayIndex))
+                .filter((topicConflict, leftExam, rightExam) -> Math.abs(leftExam.getPeriodIndex() - rightExam.getPeriodIndex()) > 1)
```
Comment (triceo) [inline optaplanner-examples/src/main/java/org/optaplanner/examples/examination/optional/score/ExaminationConstraintProvider.java]: You use this `Math.abs(...) >/= 1` construct frequently. I'd extract it to a separate method and reuse.

File: optaplanner-examples/src/main/java/org/optaplanner/examples/examination/optional/score/ExaminationConstraintProvider.java [hunk 5]
```diff
@@ -0,0 +1,371 @@
...
+                                  lessThan(Exam::getId, Exam::getId))
+                .join(Exam.class,
+                      equal(Exam::getPeriod, Exam::getPeriod),
+                      equal(Exam::getRoom, Exam::getRoom),
+                      greaterThan(Exam::getId, Exam::getId))
+                .filter((leftExam, rightExam) -> rightExam.getTopicDuration() != leftExam.getTopicDuration())
+                .ifNotExists(Exam.class,
+                             equal((leftExam, rightExam) -> rightExam.getPeriod(), Exam::getPeriod),
+                             equal((leftExam, rightExam) -> rightExam.getRoom(), Exam::getRoom),
+                             greaterThan((leftExam, rightExam) -> rightExam.getId(), Exam::getId),
```
Comment (triceo) [inline optaplanner-examples/src/main/java/org/optaplanner/examples/examination/optional/score/ExaminationConstraintProvider.java]: Personally, I like to see all `equals()` joiners first, but that's mostly aesthetic.

File: optaplanner-examples/src/main/resources/org/optaplanner/examples/examination/solver/examinationSolverConfig.xml [hunk 6]
```diff
@@ -9,6 +9,7 @@
   <entityClass>org.optaplanner.examples.examination.domain.FollowingExam</entityClass>
 
   <scoreDirectorFactory>
+<!--    <constraintProviderClass>org.optaplanner.examples.examination.optional.score.examinationConstraintProvider</constraintProviderClass>-->
```
Comment (triceo) [inline optaplanner-examples/src/main/resources/org/optaplanner/examples/examination/solver/examinationSolverConfig.xml]: Formatting, please. See other examples.

File: optaplanner-examples/src/main/resources/org/optaplanner/examples/examination/solver/examinationSolverConfig.xml [hunk 7]
```diff
@@ -9,6 +9,7 @@
...
     <scoreDrl>org/optaplanner/examples/examination/solver/examinationScoreRules.drl</scoreDrl>
```
Comment (triceo) [inline optaplanner-examples/src/main/resources/org/optaplanner/examples/examination/solver/examinationSolverConfig.xml]: Has this been through assertion score director and `FULL_ASSERT`? Since I don't see it commented out in the example, I assume not.

Do it. It will tell you whether or not your implementation is identical to DRL.

**Comment by triceo** [comment]
> I've used a lot of filtering after noticing it is not supported yet, so I've put .filter() instead.

That problem should be gone on today's master.

> * What is the ideal way how to model the != null? Two options come to mind, either use .ifNotExists() or .filter(!=null). What is the ideal one performance-wise?

I made some suggestions inline. if/when applied, I don't think there will be many `null` checks left.

> * Regarding roomCapacityTooSmall() constraint, is the drools accumulate function properly converted? If not what is the ideal way to do this?

See my comment on `FULL_ASSERT`. That will give you a very exact answer.

> * Are there any obvious performance hindering problems that I am missing?

I made some suggestions. We'll see when they're solved.

File: optaplanner-examples/src/main/java/org/optaplanner/examples/examination/optional/score/ExaminationConstraintProvider.java [hunk 8]
```diff
@@ -0,0 +1,371 @@
...
+//    $topicConflict : TopicConflict($leftTopic : leftTopic, $rightTopic : rightTopic)
+//    $leftExam : Exam(topic == $leftTopic, $period : period, period != null)
+//    $rightExam : Exam(topic == $rightTopic, period == $period)
+//    then
+//        scoreHolder.addHardConstraintMatch(kcontext, - $topicConflict.getStudentSize());
+//    end
+    private Constraint conflictingExamsInSamePeriod(ConstraintFactory constraintFactory) {
+        return constraintFactory.from(Exam.class)
+                .filter(exam -> exam.getPeriod() != null)
+                .join(Exam.class, equal(Exam::getPeriod, Exam::getPeriod))
```
Comment (triceo) [inline optaplanner-examples/src/main/java/org/optaplanner/examples/examination/optional/score/ExaminationConstraintProvider.java]: Also, these two joins can be replaced by `.fromUniquePair(Exam.class, equal(...))`

**Comment by michaltomco** [comment]
Thanks a lot for the input, I will implement them first thing tomorrow!

File: optaplanner-examples/src/main/java/org/optaplanner/examples/examination/optional/score/ExaminationConstraintProvider.java [hunk 9]
```diff
@@ -0,0 +1,371 @@
...
+//    rule "periodDurationTooShort"
+//    when
+//    $exam : Exam(topicDuration > periodDuration, period != null)
+//    then
+//        scoreHolder.addHardConstraintMatch(kcontext, - $exam.getTopicStudentSize());
+//    end
+    private Constraint periodDurationTooShort(ConstraintFactory constraintFactory) {
+        return constraintFactory.from(Exam.class)
+                .filter(exam -> exam.getTopicDuration() > exam.getPeriodDuration())
+                .ifExists(Period.class, equal(Exam::getPeriod, Function.identity()))
```
Comment (triceo) [inline optaplanner-examples/src/main/java/org/optaplanner/examples/examination/optional/score/ExaminationConstraintProvider.java]: Most likelynot necessary, as the rule doesn't use it either.

File: optaplanner-examples/src/main/java/org/optaplanner/examples/examination/optional/score/ExaminationConstraintProvider.java [hunk 10]
```diff
@@ -0,0 +1,371 @@
...
+//            Exam(period == $period, room == $room, $studentSize : topicStudentSize);
+//    $totalStudentSize : sum($studentSize);
+//    $totalStudentSize > $capacity
+//        )
+//    then
+//        scoreHolder.addHardConstraintMatch(kcontext, $capacity - $totalStudentSize);
+//    end
+    private Constraint roomCapacityTooSmall(ConstraintFactory constraintFactory) {
+        return constraintFactory.from(Exam.class)
+                .join(Period.class, equal(Exam::getPeriod, Function.identity()))
```
[the remark this answers is not in the document]
Comment (triceo) [reply optaplanner-examples/src/main/java/org/optaplanner/examples/examination/optional/score/ExaminationConstraintProvider.java]: Most likely it's a bug, though. The period is actually used in the DRL.

File: optaplanner-examples/src/main/java/org/optaplanner/examples/examination/optional/score/ExaminationConstraintProvider.java [hunk 11]
```diff
@@ -0,0 +1,371 @@
...
+//    end
+    private Constraint twoExamsInADay(ConstraintFactory constraintFactory) {
+        return constraintFactory.from(TopicConflict.class)
+                .join(Exam.class,
+                      equal(TopicConflict::getLeftTopic, Exam::getTopic))
+                .filter((topicConflict, leftExam) -> leftExam.getPeriod() != null)
+                .join(Exam.class,
+                      equal((topicConflict, leftExam) -> topicConflict.getRightTopic(), Exam::getTopic),
+                      equal((topicConflict, leftExam) -> leftExam.getDayIndex(), Exam::getDayIndex))
+                .filter((topicConflict, leftExam, rightExam) -> Math.abs(leftExam.getPeriodIndex() - rightExam.getPeriodIndex()) > 1)
```
Comment (triceo) [inline optaplanner-examples/src/main/java/org/optaplanner/examples/examination/optional/score/ExaminationConstraintProvider.java]: Although this isn't functionally equivalent to the DRL, you could get away with replacing those joins with two `ifExists()`. Good for performance, but may not pass `FULL_ASSERT`.

File: optaplanner-examples/src/main/java/org/optaplanner/examples/examination/optional/score/ExaminationConstraintProvider.java (hunk 4 above)
Comment (triceo) [reply optaplanner-examples/src/main/java/org/optaplanner/examples/examination/optional/score/ExaminationConstraintProvider.java]: Also same comment regarding `ifExists()` as with `twoExamsInADay`.

File: optaplanner-examples/src/main/java/org/optaplanner/examples/examination/optional/score/ExaminationConstraintProvider.java [hunk 12]
```diff
@@ -0,0 +1,371 @@
...
+//        scoreHolder.penalize(kcontext, $topicConflict.getStudentSize());
+//    end
+    private Constraint twoExamsInARow(ConstraintFactory constraintFactory) {
+        return constraintFactory.from(TopicConflict.class)
+                .join(Exam.class,
+                      equal(TopicConflict::getLeftTopic, Exam::getTopic))
+                .filter((topicConflict, leftExam) -> leftExam.getPeriod() != null)
+                .join(Exam.class,
+                      equal((topicConflict, leftExam) -> topicConflict.getRightTopic(), Exam::getTopic),
+                      equal((topicConflict, leftExam) -> leftExam.getDayIndex(), Exam::getDayIndex))
```
Comment (triceo) [inline optaplanner-examples/src/main/java/org/optaplanner/examples/examination/optional/score/ExaminationConstraintProvider.java]: Perhaps could be replaced by `ifExists()`, but then the constraint may no longer be functionally equivalent to the DRL. `FULL_ASSERT` would tell.

File: optaplanner-examples/src/main/java/org/optaplanner/examples/examination/optional/score/ExaminationConstraintProvider.java (hunk 11 above)
Comment (triceo) [inline optaplanner-examples/src/main/java/org/optaplanner/examples/examination/optional/score/ExaminationConstraintProvider.java]: Dtto. on `ifExists()`.

File: optaplanner-examples/src/main/java/org/optaplanner/examples/examination/optional/score/ExaminationConstraintProvider.java [hunk 13]
```diff
@@ -0,0 +1,371 @@
...
+//    // rightExam has the lowest id of the period+room+rightDuration combo
+//    not Exam(period == $period, room == $room, id < $rightId,
+//            topicDuration == $rightTopicDuration
+//    )
+//    then
+//        scoreHolder.penalize(kcontext);
+//    end
+    private Constraint mixedDurations(ConstraintFactory constraintFactory) {
+        return constraintFactory.from(Exam.class)
+                .ifExists(Exam.class, equal(Exam::getPeriod, Exam::getPeriod))
```
Comment (triceo) [inline optaplanner-examples/src/main/java/org/optaplanner/examples/examination/optional/score/ExaminationConstraintProvider.java]: Again, use the .join(Exam.class, equal(Exam::getPeriod)) shorthand.

File: optaplanner-examples/src/main/java/org/optaplanner/examples/examination/optional/score/ExaminationConstraintProvider.java [hunk 14]
```diff
@@ -0,0 +1,371 @@
...
+                .ifExists(Exam.class, equal(Exam::getRoom, Exam::getRoom))
```
Comment (triceo) [inline optaplanner-examples/src/main/java/org/optaplanner/examples/examination/optional/score/ExaminationConstraintProvider.java]: Why is this here? I don't see it in the DRL.

File: optaplanner-examples/src/main/java/org/optaplanner/examples/examination/optional/score/ExaminationConstraintProvider.java [hunk 15]
```diff
@@ -0,0 +1,371 @@
...
+                .ifNotExistsOther(Exam.class,
+                                  equal(Exam::getPeriod, Exam::getPeriod),
+                                  equal(Exam::getRoom, Exam::getRoom),
+                                  lessThan(Exam::getId, Exam::getId))
+                .join(Exam.class,
+                      equal(Exam::getPeriod, Exam::getPeriod),
+                      equal(Exam::getRoom, Exam::getRoom),
```
Comment (triceo) [inline optaplanner-examples/src/main/java/org/optaplanner/examples/examination/optional/score/ExaminationConstraintProvider.java]: Since both left and right are the same type and we're joining on the same field, `equal()` can lose its second argument.

File: optaplanner-examples/src/main/java/org/optaplanner/examples/examination/optional/score/ExaminationConstraintProvider.java [hunk 16]
```diff
@@ -0,0 +1,371 @@
...
+                      greaterThan(Exam::getId, Exam::getId))
```
Comment (triceo) [inline optaplanner-examples/src/main/java/org/optaplanner/examples/examination/optional/score/ExaminationConstraintProvider.java]: Since both left and right are the same type and we're joining on the same field, `equal()` can lose its second argument.

File: optaplanner-examples/src/main/java/org/optaplanner/examples/examination/optional/score/ExaminationConstraintProvider.java [hunk 17]
```diff
@@ -0,0 +1,371 @@
...
+
+    //// Two exams in the same period which share students.
+//    rule "conflictingExamsInSamePeriod"
+//    when
+//    $topicConflict : TopicConflict($leftTopic : leftTopic, $rightTopic : rightTopic)
+//    $leftExam : Exam(topic == $leftTopic, $period : period, period != null)
+//    $rightExam : Exam(topic == $rightTopic, period == $period)
+//    then
+//        scoreHolder.addHardConstraintMatch(kcontext, - $topicConflict.getStudentSize());
+//    end
```
[the remark this answers is not in the document]
Comment (michaltomco) [reply optaplanner-examples/src/main/java/org/optaplanner/examples/examination/optional/score/ExaminationConstraintProvider.java]: Yes, that was meant as a help for myself and potentially the reviewer. I will remove it after I fix all the issues.

**Comment by michaltomco** [comment]
I still need to debug the CS, since it returns different score values than drools.

File: optaplanner-examples/src/main/java/org/optaplanner/examples/examination/optional/score/ExaminationConstraintProvider.java [hunk 18]
```diff
@@ -0,0 +1,369 @@
...
+    private Constraint roomPenalty(ConstraintFactory constraintFactory) {
+        return constraintFactory.from(Room.class)
+                .filter(room -> room.getPenalty() != 0)
+                .ifExists(Exam.class, equal(Function.identity(), Exam::getRoom))
+                .penalize("Room penalty", HardSoftScore.ONE_SOFT, Room::getPenalty);
+    }
+
+    private int getAbsoluteValue(Exam leftExam, Exam rightExam) {
+        return Math.abs(leftExam.getPeriodIndex() - rightExam.getPeriodIndex());
+    }
```
Comment (triceo) [inline optaplanner-examples/src/main/java/org/optaplanner/examples/examination/optional/score/ExaminationConstraintProvider.java]: Choose a better name for the method. Don't describe the code, describe the purpose.

**Comment by michaltomco** [comment]
The performance is still very slow. I will let you know when I discover the culprit.

**Comment by triceo** [comment]
> The performance is still very slow. I will let you know when I discover the culprit.

I suggest an experiment, @michaltomco : remove the one constraint where you use `groupBy()`. If the performance is magically fixed, don't look any further - you've hit the CS-D performance bug and the results are as bad as expected.

File: optaplanner-examples/src/main/java/org/optaplanner/examples/examination/solver/score/ExaminationConstraintProvider.java [hunk 19]
```diff
@@ -0,0 +1,241 @@
...
+                .ifExists(Exam.class,
+                        equal((topicConflict, leftExam) -> topicConflict.getRightTopic(), Exam::getTopic),
+                        equal((topicConflict, leftExam) -> leftExam.getDayIndex(), Exam::getDayIndex),
+                        // Find exams in a day, but not being held right after each other. That case is handled in the twoExamsInARow constraint.
+                        filtering((topicConflict, leftExam,
+                                rightExam) -> getPeriodIndexDifferenceBetweenExams(leftExam, rightExam) > 1))
+                .join(ExaminationConstraintConfiguration.class)
+                .penalize("Two exams in a day",
+                        HardSoftScore.ONE_SOFT,
+                        (topicConflict, leftExam, config) -> topicConflict.getStudentSize() * config.getTwoInADayPenalty());
```
Comment (triceo) [inline optaplanner-examples/src/main/java/org/optaplanner/examples/examination/solver/score/ExaminationConstraintProvider.java:173]: I see that there is a `ConstraintConfiguration` available. Yet, you still use `penalize()` instead of `penalizeConfigurable()`. I suggest you change that.

File: optaplanner-examples/src/main/java/org/optaplanner/examples/examination/solver/score/ExaminationConstraintProvider.java [hunk 20]
```diff
@@ -0,0 +1,241 @@
...
+                        (config, topicConflict, leftExam) -> topicConflict.getStudentSize());
+    }
+
+    protected Constraint mixedDurations(ConstraintFactory constraintFactory) {
+        return constraintFactory.from(Exam.class)
+                .ifExists(Exam.class, equal(Exam::getPeriod, Exam::getPeriod))
+                .ifNotExistsOther(Exam.class,
+                        equal(Exam::getPeriod, Exam::getPeriod),
+                        equal(Exam::getRoom, Exam::getRoom),
+                        lessThan(Exam::getId, Exam::getId))
```
Comment (triceo) [inline optaplanner-examples/src/main/java/org/optaplanner/examples/examination/solver/score/ExaminationConstraintProvider.java:199]: return constraintFactory.from(Exam.class)
                .ifNotExists(Exam.class,
                        equal(Exam::getPeriod),
                        equal(Exam::getRoom),
                        lessThan(Exam::getId))

But that won't change much, if anything.

Comment (triceo) [reply optaplanner-examples/src/main/java/org/optaplanner/examples/examination/solver/score/ExaminationConstraintProvider.java:199]: Also, I'd put the `lessThan` filter first. Will eliminate many thing quickly.

File: optaplanner-examples/src/main/java/org/optaplanner/examples/examination/solver/score/ExaminationConstraintProvider.java [hunk 21]
```diff
@@ -0,0 +1,241 @@
...
+                .join(Exam.class,
+                        equal(Exam::getPeriod),
+                        equal(Exam::getRoom),
+                        greaterThan(Exam::getId),
+                        filtering((leftExam, rightExam) -> rightExam.getTopicDuration() != leftExam.getTopicDuration()))
```
Comment (triceo) [inline optaplanner-examples/src/main/java/org/optaplanner/examples/examination/solver/score/ExaminationConstraintProvider.java:204]: Also, I wonder why this entire construct isn't simply `.fromUniquePair(Exam.class, ...)`

File: optaplanner-examples/src/main/java/org/optaplanner/examples/examination/solver/score/ExaminationConstraintProvider.java [hunk 22]
```diff
@@ -0,0 +1,241 @@
...
+                .join(Exam.class,
+                        filtering((config, exam) -> exam.isTopicFrontLoadLarge() && exam.isPeriodFrontLoadLast()))
+                .penalize("Front load", HardSoftScore.ONE_SOFT, (config, exam) -> config.getFrontLoadPenalty());
+    }
+
+    protected Constraint periodPenalty(ConstraintFactory constraintFactory) {
+        return constraintFactory.from(Exam.class)
+                .ifExists(Period.class,
+                        equal(Exam::getPeriod, Function.identity()),
+                        filtering((exam, period) -> period.getPenalty() != 0))
```
Comment (triceo) [inline optaplanner-examples/src/main/java/org/optaplanner/examples/examination/solver/score/ExaminationConstraintProvider.java:226]: Why not `.filter(exam -> exam.getPeriod().getPenalty() != 0)`?

File: optaplanner-examples/src/main/java/org/optaplanner/examples/examination/solver/score/ExaminationConstraintProvider.java [hunk 23]
```diff
@@ -0,0 +1,241 @@
...
+                        equal((leftExam, rightExam) -> rightExam.getRoom(), Exam::getRoom),
+                        equal((leftExam, rightExam) -> rightExam.getTopicDuration(), Exam::getTopicDuration),
+                        greaterThan((leftExam, rightExam) -> rightExam.getId(), Exam::getId))
+                .join(ExaminationConstraintConfiguration.class)
+                .penalize("Mixed durations", HardSoftScore.ONE_SOFT,
+                        (leftExam, rightExam, config) -> config.getMixedDurationPenalty());
+    }
+
+    protected Constraint frontLoad(ConstraintFactory constraintFactory) {
+        return constraintFactory.from(ExaminationConstraintConfiguration.class)
```
Comment (triceo) [inline optaplanner-examples/src/main/java/org/optaplanner/examples/examination/solver/score/ExaminationConstraintProvider.java:216]: All uses of from/join on `ExaminationConstraintConfiguration` should theoretically be removable when converted to `penalizeConfigurable()`.

File: optaplanner-examples/src/main/java/org/optaplanner/examples/examination/solver/score/ExaminationConstraintProvider.java [hunk 24]
```diff
@@ -0,0 +1,241 @@
...
+                        equal(Exam::getPeriod, Function.identity()),
+                        filtering((exam, period) -> period.getPenalty() != 0))
+                .penalize("Period penalty", HardSoftScore.ONE_SOFT, exam -> exam.getPeriod().getPenalty());
+    }
+
+    protected Constraint roomPenalty(ConstraintFactory constraintFactory) {
+        return constraintFactory.from(Exam.class)
+                .ifExists(Room.class,
+                        equal(Exam::getRoom, Function.identity()),
+                        filtering((exam, room) -> room.getPenalty() != 0))
```
Comment (triceo) [inline optaplanner-examples/src/main/java/org/optaplanner/examples/examination/solver/score/ExaminationConstraintProvider.java:234]: Ditto. It appears you have a Room reference on `Exam` directly.

File: optaplanner-examples/src/main/java/org/optaplanner/examples/examination/solver/score/ExaminationConstraintProvider.java [hunk 25]
```diff
@@ -0,0 +1,241 @@
...
+                        equal((topicConflict, leftExam) -> topicConflict.getRightTopic(), Exam::getTopic),
+                        equal((topicConflict, leftExam) -> leftExam.getPeriod(), Exam::getPeriod))
+                .penalize("Conflicting exams in same period",
+                        HardSoftScore.ONE_HARD,
+                        (topicConflict, leftExam) -> topicConflict.getStudentSize());
+    }
+
+    protected Constraint periodDurationTooShort(ConstraintFactory constraintFactory) {
+        return constraintFactory.fromUnfiltered(Exam.class)
+                .filter(exam -> exam.getPeriod() != null)
```
Comment (triceo) [inline optaplanner-examples/src/main/java/org/optaplanner/examples/examination/solver/score/ExaminationConstraintProvider.java:65]: Why `fromUnfiltered()`, if you then filter anyway?

**Comment by triceo** [comment]
Let's not leave the constraint configuration unfinished - there should be no `penalize()` method left.
Also, when you've got all the constraints in the configuration, please make sure that the DRL is using them, too.

**Comment by michaltomco** [comment]
The performance issue causing constraints are definitely the twoExamsInARow, twoExamsInADay, and periodSpread. I'll try to rewrite them to see if it helps.

File: optaplanner-examples/src/main/resources/org/optaplanner/examples/examination/solver/examinationSolverConfig.xml [hunk 26]
```diff
@@ -1,7 +1,7 @@
 <?xml version="1.0" encoding="UTF-8"?>
 <solver>
-  <!--<environmentMode>FULL_ASSERT</environmentMode>--><!-- To slowly prove there are no bugs in this code -->
-  <moveThreadCount>AUTO</moveThreadCount><!-- To solve faster by saturating multiple CPU cores -->
+<!--  <environmentMode>FULL_ASSERT</environmentMode>&lt;!&ndash; To slowly prove there are no bugs in this code &ndash;&gt;-->
```
Comment (rsynek) [inline optaplanner-examples/src/main/resources/org/optaplanner/examples/examination/solver/examinationSolverConfig.xml:3]: It looks like accidentally wrapping one comment into another.

File: optaplanner-benchmark/pom.xml [hunk 27]
```diff
@@ -6,7 +6,7 @@
   <parent>
     <groupId>org.optaplanner</groupId>
     <artifactId>optaplanner</artifactId>
-    <version>7.39.0-SNAPSHOT</version>
+    <version>7.39.0.CR1</version>
```
[the remark this answers is not in the document]
Comment (michaltomco) [reply optaplanner-benchmark/pom.xml]: Oops, this commit must've sneaked there while rebasing. Fixed now!

File: optaplanner-examples/src/main/resources/org/optaplanner/examples/examination/solver/examinationSolverConfig.xml [hunk 28]
```diff
@@ -1,7 +1,7 @@
 <?xml version="1.0" encoding="UTF-8"?>
 <solver>
-  <!--<environmentMode>FULL_ASSERT</environmentMode>--><!-- To slowly prove there are no bugs in this code -->
-  <moveThreadCount>AUTO</moveThreadCount><!-- To solve faster by saturating multiple CPU cores -->
+<!--  <environmentMode>FULL_ASSERT</environmentMode> To slowly prove there are no bugs in this code -->
+  <!--<moveThreadCount>AUTO</moveThreadCount>--><!-- To solve faster by saturating multiple CPU cores -->
```
[the remark this answers is not in the document]
Comment (michaltomco) [reply optaplanner-examples/src/main/resources/org/optaplanner/examples/examination/solver/examinationSolverConfig.xml]: Done.

**Comment by triceo** [comment]
Thanks, glad to see this merged!
